### DISCUSSION TITLE
CONTRIBUTING.md instructions are not working for me (python newbie)

### DISCUSSION THREAD

**Discussion by plastikfan** [Textualize/rich#1994] [Q&A]
I'm just getting to grips with python but having a few teething problems with env/tools. I have just followed the instructions in the contributing file with the aim of being able to run the tests. However, when I run the `make test` command, it fails:

```
plastikfan@Janus  ~/dev/github/3rd/py/rich λ   master  make test
TERM=unknown pytest --cov-report term-missing --cov=rich tests/ -vv
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cov-report --cov=rich tests/
  inifile: /home/plastikfan/dev/github/3rd/py/rich/tests/pytest.ini
  rootdir: /home/plastikfan/dev/github/3rd/py/rich/tests

make: *** [Makefile:2: test] Error 4
```

But the next instruction says that if you don't have make (which I do), run this command:

> pytest --cov-report term-missing --cov=rich tests/ -vv

which fails again:

```
plastikfan@Janus  ~/dev/github/3rd/py/rich λ   master  pytest --cov-report term-missing --cov=rich tests/ -vv
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --cov-report --cov=rich tests/
  inifile: /home/plastikfan/dev/github/3rd/py/rich/tests/pytest.ini
  rootdir: /home/plastikfan/dev/github/3rd/py/rich/tests
```

I've spent a week or so reading up on all things python and I have made great progress, but the one area I'm a bit stuck on is generally running tasks on a python package. (Is there a standard taskrunner like concept in python, or do we have to define our own tasks with make) I'm coming from a javascript/nodejs background, where we use a package.json to define custom tasks like build, clean, test, etc, but I dont know what the standard equivalent is in python. Is make the standard way of doing this, or are there other tools too?  Its only by digging around in the markdown files that I discovered the development env in the contributing file, which pointed me in the right direction, but I'm hitting a few errors I dont know how to fix.

Another query:

I was expecting a virtual environment to be setup after installing, but there is no hint of one anywhere (.venv directory?). What am i missing?

Grateful for any help, thanks. (PS, I've run all the previous steps in the contributing file, such as installing dependencies)

**Comment by fruch** [comment]
You need todo 
pip install pytest-cov
or 
potery install

i.e. the reason you have this failure is cause you have a missing package in your environment

**Reply by plastikfan** [reply]
Ok thanks for that. I tried `poetry install`:

```
 ✘ plastikfan@Janus  ~/dev/github/3rd/py/rich λ   master  poetry install
Installing dependencies from lock file

No dependencies to install or update

Installing the current project: rich (11.2.0)
```

but that itself made no difference. Then I installed pytest-cov, then re-ran, but I get a host of other errors, too much output to requote,but analysing the output shows that there are 6 errors which look like depenency errors:

```
ModuleNotFoundError: No module named 'commonmark'
ModuleNotFoundError: No module named 'pygments'
```

This leads me to think there is still something I'm missing. Surely, when you pull down a repo, then perform install, all dependencies should be pulled down, including the pytest-cov dependency you suggested to install and then commonmark and pygments, which also appears to be missing.

What am I doing wrong? If I manually install commonmark , pygments and pytest-cov, then the tests pass, but why werent these pulled down as part of the poetry install? How would I know to install those, without having to see these errors first? Thanks

**Reply by plastikfan** [reply]
I just noticed this in the pyproject file:

```
[tool.poetry.dev-dependencies]
pytest = "^7.0.0"
black = "^22.1"
mypy = "^0.930"
pytest-cov = "^3.0.0"
attrs = "^21.4.0"
types-dataclasses = "^0.6.4"
pre-commit = "^2.17.0"
```

which makes me wonder why some of these dev dependencies (pytest-cov) were not installed, giving rise to the errors previously mentioned.

**Reply by fruch** [reply]
I think I understand, by default poetry is creating a virtualenv, and I'm guessing you are no in it.
try:
```
poetry shell
pytest
```

**Reply by plastikfan** [reply]
> I think I understand, by default poetry is creating a virtualenv, and I'm guessing you are no in it. try:
> 
> ```
> poetry shell
> pytest
> ```

Well I was wondering where the virtual environment was. I was assuming there would be a .venv directory in the root of the package, but there isnt one. Perhaps the virtualenv is in a different directory? When I run the above commands, the tests complete sucessfully, but I dont know what `poetry shell` does and how that relates to a virtual env.

**Reply by willmcgugan** [reply]
`poetry shell` creates and activates a virtual env for you.

**Reply by plastikfan** [reply]
Thanks

**Comment by willmcgugan** [comment]
CONTRIBUTING.md may be out of date, and contain too much assumed knowledge.

Will post an issue to revisit it.

**Reply by plastikfan** [reply]
thanks @willmcgugan. I appreciate you acknowledging this. I think I have to do a lot more reading up on poetry and get some more fundamental understanding of that, but I initially thought it would be relatively easy to just run tests, but I guess I need to have a bit more environment knowledge first. PS, love this package and what it provides, it's gonna save me a lot of work in the long run, because this is the type of functionality I was looking to build myself, having recently switched away from PowerShell/Javascript for cli functionality. Thanks!

**Reply by willmcgugan** [reply]
https://github.com/Textualize/rich/issues/1997
